File: liquipedia/liquipedia_test.go
package liquipedia

import (
	"os"
	"strings"
	"testing"
	"time"

	"github.com/PuerkitoBio/goquery"
	"github.com/stretchr/testify/suite"
)

type LiquipediaTestSuite struct {
	suite.Suite

	Brame                   *goquery.Selection
	Creepwave               *goquery.Selection
	ThePrimeAndArmyGeniuses *goquery.Selection
}

func TestLiquipediaTestSuite(t *testing.T) {
	suite.Run(t, new(LiquipediaTestSuite))
}

func (s *LiquipediaTestSuite) SetupTest() {
	file, err := os.Open("testdata/brame.html")
	s.Nil(err)
	doc, err := goquery.NewDocumentFromReader(file)
	s.Nil(err)
	s.Brame = doc.Children()

	file, err = os.Open("testdata/creepwave.html")
	s.Nil(err)
	doc, err = goquery.NewDocumentFromReader(file)
	s.Nil(err)
	s.Creepwave = doc.Children()

	file, err = os.Open("testdata/the-prime-and-army-geniuses.html")
	s.Nil(err)
	doc, err = goquery.NewDocumentFromReader(file)
	s.Nil(err)
	s.ThePrimeAndArmyGeniuses = doc.Children()
}

func (s *LiquipediaTestSuite) TearDownTest() {
	s.Brame = nil
	s.Creepwave = nil
	s.ThePrimeAndArmyGeniuses = nil
}

func (s *LiquipediaTestSuite) TestDescriptionLinks() {
	table := []struct {
		in   *goquery.Selection
		link string
	}{
		{
			in:   s.Brame,
			link: "<a href=\"/dota2/Brame\" title=\"Brame\">Brame</a>",
		},
		{
			in:   s.Creepwave,
			link: "<a href=\"/dota2/Creepwave\" title=\"Creepwave\">Creepwave</a>",
		},
		{
			in:   s.ThePrimeAndArmyGeniuses,
			link: "<a href=\"/dota2/The_Prime\" title=\"The Prime\">The Prime</a>",
		},
		{
			in:   s.ThePrimeAndArmyGeniuses,
			link: "<a href=\"/dota2/Army_Geniuses\" title=\"Army Geniuses\">Army Geniuses</a>",
		},
	}

	for _, tt := range table {
		desc, err := Description(tt.in)
		s.Nil(err)
		s.Contains(desc, tt.link)
		s.Equal(1, strings.Count(desc, tt.link))
	}
}

func (s *LiquipediaTestSuite) TestDescriptionRemovedFlags() {
	table := []struct {
		in    *goquery.Selection
		flags []string
	}{
		{
			in: s.Brame,
			flags: []string{
				"Ukraine",
			},
		},
		{
			in: s.Creepwave,
			flags: []string{
				"Belarus",
				"Bulgaria",
				"Jordan",
				"Netherlands",
				"Russia",
			},
		},
		{
			in: s.ThePrimeAndArmyGeniuses,
			flags: []string{
				"Indonesia",
			},
		},
	}

	for _, tt := range table {
		desc, err := Description(tt.in)
		s.Nil(err)
		for _, flag := range tt.flags {
			s.NotContains(desc, flag)
		}
	}
}

func (s *LiquipediaTestSuite) TestDescriptionRemovedImages() {
	list := []*goquery.Selection{
		s.Brame,
		s.Creepwave,
		s.ThePrimeAndArmyGeniuses,
	}

	for _, sel := range list {
		desc, err := Description(sel)
		s.Nil(err)
		s.NotEmpty(desc)

		doc, err := goquery.NewDocumentFromReader(strings.NewReader(desc))
		s.Nil(err)
		s.Zero(doc.Find("img").Length())
	}
}

func (s *LiquipediaTestSuite) TestDescriptionRemovedRef() {
	list := []*goquery.Selection{
		s.Brame,
		s.Creepwave,
		s.ThePrimeAndArmyGeniuses,
	}

	for _, sel := range list {
		desc, err := DescriptionWithoutRef(sel)
		s.Nil(err)
		s.NotEmpty(desc)

		doc, err := goquery.NewDocumentFromReader(strings.NewReader(desc))
		s.Nil(err)
		s.Zero(doc.Find(".Ref").Length())
	}
}

func (s *LiquipediaTestSuite) TestID() {
	date := time.Date(2021, time.October, 1, 12, 0, 0, 0, time.UTC)
	id := ID(date, "example text")
	s.Equal("tag:liquipedia.net,2021-10-01:DpSuNtpv8DmSpX/dvfRyi2CdDX/m6wGfqfG5tbVA2DU", id)
}

func (s *LiquipediaTestSuite) TestTitle() {
	table := []struct {
		in    *goquery.Selection
		title string
	}{
		{
			in:    s.Brame,
			title: "Nefrit",
		},
		{
			in:    s.Creepwave,
			title: "ATF Chu Crystallis Fishman hansha",
		},
		{
			in:    s.ThePrimeAndArmyGeniuses,
			title: "Azur4",
		},
	}

	for _, tt := range table {
		// Ensure 'Description()' has no side-effects
		Description(tt.in)
		s.Equal(tt.title, Title(tt.in))
	}
}

File: src/liquipedia.test.ts
import assert from 'assert';
import path from 'path';
import { readFileSync } from 'fs';
import { load } from 'cheerio';
import { description, descriptionWithoutRef, entryID } from './liquipedia';

const Brame = 'Brame';
const BrameDiffRef = 'BrameDiffRef';
const Creepwave = 'Creepwave';
const ThePrimeAndArmyGeniuses = 'ThePrimeAndArmyGeniuses';

type TeamData = {
  [key: string]: string;
};

async function loadTestData(): Promise<TeamData> {
  const data = {
    Brame: await readFile('brame.html'),
    BrameDiffRef: await readFile('brame-diff-ref.html'),
    Creepwave: await readFile('creepwave.html'),
    ThePrimeAndArmyGeniuses: await readFile('the-prime-and-army-geniuses.html'),
  };
  return data;
}

async function readFile(name: string): Promise<string> {
  const file = path.join(__dirname, './testdata/', name);
  const content = readFileSync(file);
  return content.toString();
}

it('description contains a single link for each team', async function () {
  const data = await loadTestData();
  const tests = [
    {
      in: Brame,
      href: '/dota2/Brame',
    },
    {
      in: Creepwave,
      href: '/dota2/Creepwave',
    },
    {
      in: ThePrimeAndArmyGeniuses,
      href: '/dota2/The_Prime',
    },
    {
      in: ThePrimeAndArmyGeniuses,
      href: '/dota2/Army_Geniuses',
    },
  ];

  for (const t of tests) {
    const desc = description(data[t.in]);
    const $ = load(desc);
    const filter = `a[href*="${t.href}"]`;
    assert.equal($(filter).length, 1);
  }
});

it('description has no flags', async function () {
  const data = await loadTestData();
  const tests = [
    {
      in: Brame,
      flags: ['Ukraine'],
    },
    {
      in: Creepwave,
      flags: ['Belarus', 'Bulgaria', 'Jordan', 'Netherlands', 'Russia'],
    },
    {
      in: ThePrimeAndArmyGeniuses,
      flags: ['Indonesia'],
    },
  ];

  for (const t of tests) {
    const desc = description(data[t.in]);
    for (const f of t.flags) {
      assert.equal(desc.includes(f), false);
    }
  }
});

it('description has no images', async function () {
  const data = await loadTestData();
  const tests = [Brame, Creepwave, ThePrimeAndArmyGeniuses];

  for (const t of tests) {
    const desc = description(data[t]);
    const $ = load(desc);
    assert.equal($('img').length, 0);
  }
});

it('description used for hashing has no ref', async function () {
  const data = await loadTestData();
  const tests = [Brame, Creepwave, ThePrimeAndArmyGeniuses];

  for (const t of tests) {
    const desc = descriptionWithoutRef(data[t]);
    const $ = load(desc);
    assert.equal($('.Ref').length, 0);
  }
});

it('entry id contains hash', async function () {
  const data = await loadTestData();
  const tests = [
    {
      in: Brame,
      id: 'tag:liquipedia.net,2021-10-21:zHB8tgNDBQKQa6gbfhps9LY6zThmlZJN6pon3T9Xp04=',
    },
    {
      in: Creepwave,
      id: 'tag:liquipedia.net,2021-10-22:loV/ga28Pof6XJqRv28KcrjD+NXW+uKkPC6aQFW0eck=',
    },
    {
      in: ThePrimeAndArmyGeniuses,
      id: 'tag:liquipedia.net,2021-10-22:cOPmYZxq/B+/s/nAipZ4zPguIhbiDrzvVkUfX8Svhy8=',
    },
  ];

  for (const t of tests) {
    const id = await entryID(data[t.in]);
    assert.equal(t.id, id);
  }
});

it('entry id does not depend on ref', async function () {
  const brameID =
    'tag:liquipedia.net,2021-10-21:zHB8tgNDBQKQa6gbfhps9LY6zThmlZJN6pon3T9Xp04=';
  const data = await loadTestData();
  const tests = [Brame, BrameDiffRef];

  for (const t of tests) {
    const id = await entryID(data[t]);
    assert.equal(brameID, id);
  }
});
